src/macros/macrorecorder.cpp

```diff
@@ -0,0 +1,106 @@
+#include "macrorecorder.h"
+
+#include <QtConcurrentRun>
+
+#include "control/controlproxy.h"
+#include "preferences/configobject.h"
+
+namespace {
+constexpr uint kMaxMacroSize = 1000;
+const QString kConfigGroup = QStringLiteral("[MacroRecording]");
+}
+
+MacroRecorder::MacroRecorder()
+        : m_COToggleRecording(ConfigKey(kConfigGroup, "recording_toggle")),
+          m_CORecStatus(ConfigKey(kConfigGroup, "recording_status")),
+          m_activeChannel(nullptr),
+          m_pStartRecordingTimer(this),
+          m_pRecordedActions(kMaxMacroSize) {
+    qCDebug(macroLoggingCategory) << "MacroRecorder construct";
+
+    connect(&m_COToggleRecording,
+            &ControlPushButton::valueChanged,
+            this,
+            &MacroRecorder::slotToggleRecording);
+    connect(&m_pStartRecordingTimer,
+            &QTimer::timeout,
+            this,
+            &MacroRecorder::pollRecordingStart);
+}
+
+void MacroRecorder::notifyCueJump(
+        ChannelHandle* channel, double sourceFramePos, double destFramePos) {
+    qCDebug(macroLoggingCategory) << "Jump in channel" << channel->handle();
+    if (isRecordingActive() && checkOrClaimRecording(channel)) {
+        m_pRecordedActions.emplace(sourceFramePos, destFramePos);
```

Comment:
emplace() will enter a hot loop if the queue is full and the real-time thread gets stuck. Use try_emplace().